Onboarding Improvements: Epilogue Screen Select Site

ParaskP7 opened the pull request:
Parent #14990

This PR enables selecting a site from within the newly designed `Login Epilogue` screen.

NOTE: Selecting a site from the newly design `Login Epilogue` screen will now always show the `Quick Start Prompt`, no matter whether the user selected `Show me around` or `No thanks`. So, if the user logs-out and logs-back-in afterwards, then selects the very same site, then the `Quick Start Prompt` will be shown again.

-----

To test:

Scenario: `Dismiss`
1. Go to `My Site` -> `App Settings` -> `Test feature configuration`.
2. Enable the `OnboardingImprovementsFeatureConfig` feature flag.
3. Log-out and log-in to the app again.
4. When on the `Login Epilogue` screen, select a site from the list of sites.
5. Make sure the new `Quick Start Prompt` is shown.
6. Dismiss the fullscreen dialog by clicking or swiping back.
7. Make sure the `My Site` screen is show and that the `Quick Start` section, with `Next Steps` and all is not shown. 

Scenario: `No thanks`
1. Go to `My Site` -> `App Settings` -> `Test feature configuration`.
2. Enable the `OnboardingImprovementsFeatureConfig` feature flag.
3. Log-out and log-in to the app again.
4. When on the `Login Epilogue` screen, select a site from the list of sites.
5. Make sure the new `Quick Start Prompt` is shown.
6. Click the `No thanks` button.
7. Make sure the `My Site` screen is shown and that the `Quick Start` section, with its `Next Steps` and all, is not shown.

Scenario: `Show me around`
1. Go to `My Site` -> `App Settings` -> `Test feature configuration`.
2. Enable the `OnboardingImprovementsFeatureConfig` feature flag.
3. Log-out and log-in to the app again.
4. When on the `Login Epilogue` screen, select a site from the list of sites.
5. Make sure the new `Quick Start Prompt` is shown.
6. Click the `Show me around` button.
7. Make sure the `My Site` screen is shown and that the `Quick Start` section, with its `Next Steps` and all, is shown.

-----

## Regression Notes
1. Potential unintended areas of impact

The existing `Quick Start` logic might be somehow affected by all the changes in this PR since there has been some updates to all that.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

See `To test` section above.

3. What automated tests I added (or what prevented me from doing so)

N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ParaskP7 committed:
Refactor: Make site picker adapter's on site click listener public

This refactor is required in order to enabled setting the
'OnSiteClickListener' from within the 'LoginEpilogueFragment' class.

ParaskP7 committed:
Feature: Add new on site click method to login epilogue listener

This newly added method will used from within the
'LoginEpilogueFragment' class and the 'OnSiteClickListener' listener to
push the on site click event into from the fragment and into the
activity layer.

ParaskP7 committed:
Feature: Set on site click listener on the login epilogue adapter

The 'OnSiteClickListener' listener will be only set when the
'Onboarding Improvements' feature flag is turned on, otherwise, the
'onSiteClick(...)' even will be ignored.

ParaskP7 committed:
Feature: Add new select site navigation event

This navigation event will be used to notify the activity that a site
has been selected so that it can then act on it accordingly.

ParaskP7 committed:
Test: Add tests for the on site click login epilogue view model action

These tests will fail as the implementation is not ready yet. The next
commit will make the tests pass (TDD).

ParaskP7 committed:
Feature: Implement on site click action

This commit makes the previous failing test pass (TDD).

ParaskP7 committed:
Feature: Enable clicking on the last item of the site picker adapter

When the 'Onboarding Improvmenents' feature flag is enabled, the last
item of the 'SitePickerAdapter' is no longer the 'Connect another site'
but instead an actual site. As such, this site needs to be clickable
just like all the other sites in the list.

ParaskP7 committed:
Feature: Launch show login epilogue with start activity for result

This upgrade is necessary in order for 'WPMainActivity' to be able to
receive the select site trigger from the 'LoginEpilogueActivity' and
deal with that event accordingly, through the newly added
'RequestCodes.LOGIN_EPILOGUE', just like it is already done for the
existing 'RequestCodes.SITE_PICKER' logic.

ParaskP7 committed:
Feature: Implement select site functionality from the login epilogue

ParaskP7 committed:
Feature: Add is quick start done util wrapper function

This newly added function, in addition to the already existing
'isQuickStartInProgress(...)' fucntion, will help decide whether the
quick start prompt is to be shown when a site is being selected from
within the login epilogue screen.

The 'isQuickStartInProgress(...)' will help identify whether the quick
start is in progress, and if so, the quick start prompt will not show.

The 'isQuickStartDone(...)' will help identify whether the quick start
is done, and if so, the quick start prompt will not show.

As such, the quick start prompt will only shown when the quick start is
not in progress and it is not already done, thus achieving the decired
logic of showing the quick start to already existing users or new ones.

ParaskP7 committed:
Feature: Show quick start prompt when not already in progress or done

ParaskP7 committed:
Feature: Skip quick start indefinitely upon user selecting no thanks

When the user is prompted with the quick start dialog, they can decide
to:
- 'Show me around': Which enables the quick start.
- 'No thanks': Which will now disable the quick start by marking all the
steps as done.
- 'Dismiss': Which ignores the dialog for now, but it will show up again
next time.

ashiagr commented:
👋 @ParaskP7,

Thanks for asking for feedback on this draft PR. 🙏

I like how you've split every step into separate commits and created a small, focused PR to make reviewing easy for a task that depends on several if conditions. I'm glad we decided on this approach as I was actually able to find something we might want to discuss:

We used local site id to check if the site has quick start steps in progress in commit 75194e9. 
It seems that every time we re-login, the local site id doesn't remain the same. Because of this, if we select "No thanks" in the quick start prompt dialog, it will have no effect after we re-login and re-select the site i.e. the quick start prompt dialog will be re-shown for the same site in spite of the  "No thanks" selection. Is this expected?

Considering the PR is still in draft, just want to highlight and discuss it before we go too far with the changes.

ParaskP7 commented:
👋 @ashiagr !

> Thanks for asking for feedback on this draft PR. 🙏

Thanks for going through with reviewing this draft PR. 🙏 

> I like how you've split every step into separate commits and created a small, focused PR to make reviewing easy for a task that depends on several if conditions. I'm glad we decided on this approach as I was actually able to find something we might want to discuss:

🙏 

> We used local site id to check if the site has quick start steps in progress in commit 75194e9.

True.

> It seems that every time we re-login, the local site id doesn't remain the same. Because of this, if we select "No thanks" in the quick start prompt dialog, it will have no effect after we re-login and re-select the site i.e. the quick start prompt dialog will be re-shown for the same site in spite of the "No thanks" selection. Is this expected?

You are totally right Ashita, actually no matter what you select, `Show me around` or `No thanks` when you logout and login again, the `showQuickStartDialog()` will be re-triggered as both checks, the `isQuickStartInProgress(...)` and `isQuickStartDone(...)` will give a false negative. It indeed seems that this is because the `localSiteId` as you noted. Thanks so much for catching this!

In order for me to avoid logging out and in again all the time, while hitting the login limit while doing so, I manually updated bits of the code to always show the login epilogue screen on app start-up. Thus, I based my solution on what the site picker screen was doing (see `SitePickerActivity.selectSiteAndFinish(...)` thinking that this is the right way. However, and as it seems this is not working with what we are trying to do.

Let me figure this out and I'll come back with an alternative solution. 🙏 

> Considering the PR is still in draft, just want to highlight and discuss it before we go too far with the changes.

Definitely, let's discuss, but it is pretty obvious that we cannot use the `localSiteId` to check on the quick start state and as such I'll need to find out another way. Feel free to suggest any alternatives that you could think of so far. 🙏 

-----

Once more, this is such a nice find, thanks so much Ashita! 🙇

zwarm commented:
@ParaskP7 - I agree with @ashiagr above. Nicely broken down and easy to follow. I almost feel like should revisit the "clearing" of QS on logout or decide if this is important enough to maintain on the backend. I have felt frustration when the steps no longer appear.

ParaskP7 commented:
👋 @zwarm !

> @ParaskP7 - I agree with @ashiagr above. Nicely broken down and easy to follow.

Thanks for going through with reviewing this draft PR. 🙏

> I almost feel like should revisit the "clearing" of QS on logout or decide if this is important enough to maintain on the backend.

Hmm, yeap, and based on @leandroalonso 's input, it is completely ok for us to always display the `Quick Start Prompt` when the user logs-out and logs-in again. As such, the "clearing" of QA on logout is what we are going towards with the commit update I am going to add to this PR soonish.

> I have felt frustration when the steps no longer appear.

True.

ParaskP7 committed:
Feature: Show quick start dialog everytime

After some thorough conversation is has been decided to always show the
quick start dialog, when coming from the login epilogue screen.

This is done because the current solution anyways doesn't support saving
the quick start state during re-login, which would enable knowing
whether the quick start is in progress/done for a specific site.

The current solution is based on the 'Local Site Id', which changes for
a specific site after a re-login. As such, any quick start state that
was based on that is being unmapped during re-login and then stays
unmapped for ever.

In order to overcome this problem, the current solution needs to be
migrated to another solution, one that is based on the '(Remote) Site
Id' and 'Self Hosted Site Id' instead. However, this is a major task by
itself and as such dealt with accordingly if need be.

ParaskP7 committed:
Refactor: Use site picker activity key local id key on login epilogue

This is done in order to make sure that the same
'SitePickerActivity.KEY_LOCAL_ID' key is used both during the write and
read operation on that specific flow.

Since the read happens during 'WPMainActivity.setSite(...)' and the key
that is used there is the 'SitePickerActivity.KEY_LOCAL_ID' then it
makes sense to use the same key on the 'LoginEpilogueActivity' as well
and avoid duplication, which might cause sync problems afterwards.

ParaskP7 committed:
Feature: Do not skip quick start when no thanks option is chosen

After some thorough conversation is has been decided to always show the
quick start dialog, when coming from the login epilogue screen. As such,
clicking on the 'No thanks' option doesn't need to alter any quick start
state and can default to only triggering a track event, exactly as it
used to behave.

This is done because the current solution anyways doesn't support saving
the quick start state during re-login, which would enable knowing
whether the quick start is in progress/done for a specific site.

The current solution is based on the 'Local Site Id', which changes for
a specific site after a re-login. As such, any quick start state that
was based on that is being unmapped during re-login and then stays
unmapped for ever.

In order to overcome this problem, the current solution needs to be
migrated to another solution, one that is based on the '(Remote) Site
Id' and 'Self Hosted Site Id' instead. However, this is a major task by
itself and as such dealt with accordingly if need be.

ParaskP7 committed:
Refactor: Remove unused is quick start done utility funtion

Since it has been decided to always show the quick start dialog, this
utility function is no longer necessary.

ParaskP7 commented:
👋 @ashiagr @zwarm !

This PR is now ready for ready.

PS: Please check the commits that I just added. They are all based on the decision that was made where we will not be holding any `Quick Start` state on re-login and keep it simple for now.
PSS: I also added the `Preserve Quick Start state during re-login.` nice to have task on the parent ticket for us to revisit all that if the time allows.

ashiagr reviewed (approved):
Thank you for the discussion (internal reference: p1627036988014500-slack-C027K4MNPGQ) and for keeping everyone involved in the loop 🙌.
Also, good job including detailed notes in the commits.

I've tested the changes, everything looks good to me 👍.

> I also added the Preserve Quick Start state during re-login. nice to have task on the parent ticket for us to revisit all that if the time allows.

👍

ParaskP7 commented:
👋 @ashiagr !

> Thank you for the discussion (internal reference: p1627036988014500-slack-C027K4MNPGQ) and for keeping everyone involved in the loop 🙌.
> Also, good job including detailed notes in the commits.

Thank you for actually catching that and spinning this off. 🥇 

> I've tested the changes, everything looks good to me 👍.

Great, I am going to merge that now then, and once again thanks for doing another code review on that. 🚀
